File: Packages/TypeEngine/Sources/TypeEngine/DictLexicon.swift
import Lexicon

/// Dictionary-backed `Lexicon` for tests and the micro-eval harness.
///
/// The canonical `Lexicon` protocol lives in the Lexicon package (alongside
/// the production `FrequencyLexicon` mmap reader over `.lex` files); this
/// test double stays here because TypeEngine's tests and `type-eval` seed it
/// with tiny in-memory wordlists. Not intended for production use
/// (linear-scan completions).
public struct DictLexicon: Lexicon {
    private let unigrams: [String: UInt32]
    private let bigrams: [String: UInt32]
    public let totalUnigramTokens: UInt64

    /// - Parameters:
    ///   - unigrams: word -> frequency (words should be lowercased)
    ///   - bigrams: "first second" (single space) -> frequency
    public init(unigrams: [String: UInt32], bigrams: [String: UInt32] = [:]) {
        self.unigrams = unigrams
        self.bigrams = bigrams
        self.totalUnigramTokens = unigrams.values.reduce(0) { $0 + UInt64($1) }
    }

    public func frequency(of word: String) -> UInt32? {
        unigrams[word]
    }

    public func bigramFrequency(_ first: String, _ second: String) -> UInt32? {
        bigrams["\(first) \(second)"]
    }

    public func completions(of prefix: String, limit: Int) -> [(word: String, frequency: UInt32)] {
        guard limit > 0 else { return [] }
        return unigrams
            .filter { $0.key.hasPrefix(prefix) }
            .sorted { $0.value > $1.value || ($0.value == $1.value && $0.key < $1.key) }
            .prefix(limit)
            .map { (word: $0.key, frequency: $0.value) }
    }

    /// Bigram followers of `word`, descending bigram frequency (linear scan
    /// over the bigram table — test-double quality, mirrors
    /// `FrequencyLexicon.continuations(of:limit:)` semantics).
    public func continuations(of word: String, limit: Int) -> [(word: String, frequency: UInt32)] {
        guard limit > 0 else { return [] }
        let prefix = "\(word) "
        return bigrams
            .compactMap { key, freq -> (word: String, frequency: UInt32)? in
                guard key.hasPrefix(prefix) else { return nil }
                return (word: String(key.dropFirst(prefix.count)), frequency: freq)
            }
            .sorted { $0.frequency > $1.frequency || ($0.frequency == $1.frequency && $0.word < $1.word) }
            .prefix(limit)
            .map { $0 }
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/DictLexicon.kt
package `is`.solberg.lyklabord.engine

import `is`.solberg.lyklabord.engine.lexicon.LexEntry
import `is`.solberg.lyklabord.engine.lexicon.Lexicon
import `is`.solberg.lyklabord.engine.lexicon.LexiconPrefixCursor
import `is`.solberg.lyklabord.engine.lexicon.PrefixChild
import `is`.solberg.lyklabord.engine.lexicon.PrefixSearchableLexicon

/**
 * Dictionary-backed [Lexicon] for tests and the micro-eval harness.
 *
 * This is an in-memory test double, not a production lexicon reader. Its
 * completion and continuation queries intentionally use linear scans.
 *
 * 1:1 port of `Packages/TypeEngine/Sources/TypeEngine/DictLexicon.swift`.
 */
class DictLexicon(
    unigrams: Map<String, UInt>,
    bigrams: Map<String, UInt> = emptyMap(),
) : PrefixSearchableLexicon {
    private val unigrams: Map<String, UInt> = unigrams
    private val bigrams: Map<String, UInt> = bigrams

    override val totalUnigramTokens: ULong =
        unigrams.values.fold(0UL) { total, frequency -> total + frequency.toULong() }

    /** Words sorted by raw UTF-8 bytes, matching the production `.lex` pool. */
    private val sortedEntries: List<SortedEntry> = unigrams.entries
        .map { entry ->
            SortedEntry(
                bytes = entry.key.toByteArray(Charsets.UTF_8),
                word = entry.key,
                frequency = entry.value,
            )
        }
        .sortedWith { lhs, rhs -> compareUnsignedBytes(lhs.bytes, rhs.bytes) }

    private data class SortedEntry(
        val bytes: ByteArray,
        val word: String,
        val frequency: UInt,
    )

    override fun frequency(of: String): UInt? = unigrams[of]

    override fun bigramFrequency(first: String, second: String): UInt? =
        bigrams["$first $second"]

    override fun completions(of: String, limit: Int): List<LexEntry> {
        if (limit <= 0) return emptyList()
        return unigrams.asSequence()
            .filter { (word, _) -> word.startsWith(of) }
            .map { (word, frequency) -> LexEntry(word, frequency) }
            .sortedWith(FREQUENCY_DESC_THEN_WORD)
            .take(limit)
            .toList()
    }

    /**
     * Bigram followers of [of], descending bigram frequency. This is a linear
     * scan over the bigram table, mirroring `FrequencyLexicon` semantics.
     */
    override fun continuations(of: String, limit: Int): List<LexEntry> {
        if (limit <= 0) return emptyList()
        val prefix = "$of "
        return bigrams.asSequence()
            .mapNotNull { (key, frequency) ->
                if (!key.startsWith(prefix)) return@mapNotNull null
                LexEntry(key.removePrefix(prefix), frequency)
            }
            .sortedWith(FREQUENCY_DESC_THEN_WORD)
            .take(limit)
            .toList()
    }

    override fun prefixRootCursor(): LexiconPrefixCursor =
        LexiconPrefixCursor(
            lowerBound = 0,
            upperBound = sortedEntries.size,
            byteDepth = 0,
        )

    override fun descend(
        cursor: LexiconPrefixCursor,
        bytes: ByteArray,
    ): LexiconPrefixCursor {
        val depth = cursor.byteDepth + bytes.size
        if (cursor.isEmpty || bytes.isEmpty()) {
            return LexiconPrefixCursor(
                lowerBound = cursor.lowerBound,
                upperBound = cursor.lowerBound,
                byteDepth = depth,
            )
        }

        fun compare(index: Int): Int {
            val word = sortedEntries[index].bytes
            var i = 0
            while (i < bytes.size) {
                val j = cursor.byteDepth + i
                if (j >= word.size) return -1
                val wordByte = word[j].toInt() and 0xFF
                val queryByte = bytes[i].toInt() and 0xFF
                if (wordByte != queryByte) return if (wordByte < queryByte) -1 else 1
                i += 1
            }
            return 0
        }

        fun bound(strict: Boolean): Int {
            var lo = cursor.lowerBound
            var hi = cursor.upperBound
            while (lo < hi) {
                val mid = (lo + hi) shr 1
                val comparison = compare(mid)
                if (if (strict) comparison > 0 else comparison >= 0) {
                    hi = mid
                } else {
                    lo = mid + 1
                }
            }
            return lo
        }

        return LexiconPrefixCursor(
            lowerBound = bound(strict = false),
            upperBound = bound(strict = true),
            byteDepth = depth,
        )
    }

    override fun exactEntry(cursor: LexiconPrefixCursor): LexEntry? {
        if (cursor.isEmpty || cursor.byteDepth <= 0) return null
        val entry = sortedEntries[cursor.lowerBound]
        if (entry.bytes.size != cursor.byteDepth) return null
        return LexEntry(entry.word, entry.frequency)
    }

    override fun childCursors(
        cursor: LexiconPrefixCursor,
        scanLimit: Int,
    ): List<PrefixChild>? {
        if (cursor.count > scanLimit) return null
        if (cursor.isEmpty) return emptyList()

        val children = ArrayList<PrefixChild>()
        var groupStart = -1
        var groupBytes = ByteArray(0)

        fun closeGroup(endingAt: Int) {
            if (groupStart < 0) return
            val character = String(groupBytes, Charsets.UTF_8).firstOrNull() ?: return
            children += PrefixChild(
                character = character,
                cursor = LexiconPrefixCursor(
                    lowerBound = groupStart,
                    upperBound = endingAt,
                    byteDepth = cursor.byteDepth + groupBytes.size,
                ),
            )
        }

        for (index in cursor.lowerBound until cursor.upperBound) {
            val word = sortedEntries[index].bytes
            if (word.size <= cursor.byteDepth) continue

            val lead = word[cursor.byteDepth].toInt() and 0xFF
            val scalarLength = when {
                lead < 0x80 -> 1
                lead < 0xE0 -> 2
                lead < 0xF0 -> 3
                else -> 4
            }
            val end = minOf(cursor.byteDepth + scalarLength, word.size)
            val bytes = word.copyOfRange(cursor.byteDepth, end)
            if (!bytes.contentEquals(groupBytes)) {
                closeGroup(endingAt = index)
                groupStart = index
                groupBytes = bytes
            }
        }
        closeGroup(endingAt = cursor.upperBound)
        return children
    }

    private companion object {
        val FREQUENCY_DESC_THEN_WORD: Comparator<LexEntry> =
            Comparator { lhs, rhs ->
                when {
                    lhs.frequency != rhs.frequency -> rhs.frequency.compareTo(lhs.frequency)
                    else -> lhs.word.compareTo(rhs.word)
                }
            }
    }
}

private fun compareUnsignedBytes(lhs: ByteArray, rhs: ByteArray): Int {
    val n = minOf(lhs.size, rhs.size)
    var i = 0
    while (i < n) {
        val left = lhs[i].toInt() and 0xFF
        val right = rhs[i].toInt() and 0xFF
        if (left != right) return if (left < right) -1 else 1
        i += 1
    }
    return lhs.size.compareTo(rhs.size)
}
